### PR TITLE
Adding support to audio, video, PDF file inputs for multimodal models

### LINKED ISSUE #1447: [BUG] EmbeddingModelTextClassifier should batch embedding calculations
**Describe the bug**

_This is more of an improvement than a bug._

Currently, `EmbeddingModelTextClassifier` calculates the vector embeddings for each sample of each label one at a time.
This can be particularly slow when calling a cloud-based embedding model.
The idea would be to use the embedding model's `embedAll()` method, instead of `embed()` to calculate all embeddings for each label at once.

**Please complete the following information:**
- LangChain4j version: 0.32.0
- LLM(s) used: any
- Java version: any

### PR DESCRIPTION
**Pull request by glaforge**
**#1463 adding audio, video, rich format support for multimodal models**

### CODE REVIEW

File: langchain4j-core/src/main/java/dev/langchain4j/data/audio/Audio.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(url, base64Data, mimeType);
+    }
+
+    @Override
+    public String toString() {
+        return "audio {" +
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/audio/Audio.java]:
```suggestion
        return "Audio {" +
```

File: langchain4j-core/src/main/java/dev/langchain4j/data/message/AudioContent.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        return new AudioContent(base64Data, mimeType);
+    }
+
+    /**
+     * Create a new {@link AudioContent} from the given Audio.
+     *
+     * @param Audio the Audio.
+     * @return the new {@link AudioContent}.
+     */
+    public static AudioContent from(Audio Audio) {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/message/AudioContent.java]:
```suggestion
    public static AudioContent from(Audio audio) {
```

File: langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java [hunk 3]
```diff
@@ -143,6 +180,32 @@ static Part map(ImageContent content) {
...
     }
 
+    static Part map(AudioContent content) {
+        Audio audio = content.audio();
+        if (audio.url() != null) {
+            String mimeType = getOrDefault(audio.mimeType(), () -> detectMimeType(audio.url()));
+            if (audio.url().getScheme().equals("gs")) {
+                return fromMimeTypeAndData(mimeType, audio.url());
+            } else {
+                return fromMimeTypeAndData(mimeType, readBytes(audio.url().toString()));
```
Comment (langchain4j) [inline langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java]: would be nice to have a test for this case

File: langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java [hunk 4]
```diff
@@ -143,6 +180,32 @@ static Part map(ImageContent content) {
...
+            }
+        }
+        return fromMimeTypeAndData(audio.mimeType(), Base64.getDecoder().decode(audio.base64Data()));
```
Comment (langchain4j) [inline langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java]: would be nice to have a test for this case

File: langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java [hunk 5]
```diff
@@ -143,6 +180,32 @@ static Part map(ImageContent content) {
...
+    }
+
+    static Part map(VideoContent content) {
+        Video video = content.video();
+        if (video.url() != null) {
+            String mimeType = getOrDefault(video.mimeType(), () -> detectMimeType(video.url()));
+            if (video.url().getScheme().equals("gs")) {
+                return fromMimeTypeAndData(mimeType, video.url());
+            } else {
+                return fromMimeTypeAndData(mimeType, readBytes(video.url().toString()));
```
Comment (langchain4j) [inline langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java]: would be nice to have a test for this case

File: langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java [hunk 6]
```diff
@@ -143,6 +180,32 @@ static Part map(ImageContent content) {
...
+            }
+        }
+        return fromMimeTypeAndData(video.mimeType(), Base64.getDecoder().decode(video.base64Data()));
```
Comment (langchain4j) [inline langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java]: would be nice to have a test for this case

**Review by ghost — COMMENTED** [review]
@glaforge looks great, thank you!

File: langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java (hunk 3 above)
Comment (glaforge) [reply langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java]: I only tested the `gs` path and will add a case testing the URL path as well.

I used a public file we use in our samples, so it should be there for the foreseeable future, but we could think of adding some files in a cloud storage bucket in the project used by the CI/CD as well.

I'm also refactoring the various `map()` methods as they are mostly duplicated code. That way, the test will exercise both paths more.

Comment (glaforge) [reply langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java]: Refactored, added a bit of test coverage.
Both gs and https path should be covered.

File: langchain4j-core/src/main/java/dev/langchain4j/data/message/ContentType.java [hunk 7]
```diff
@@ -12,7 +12,19 @@ public enum ContentType {
     /**
      * Image content.
      */
-    IMAGE(ImageContent.class);
+    IMAGE(ImageContent.class),
+    /**
+     * Audio content.
+     */
+    AUDIO(AudioContent.class),
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/message/ContentType.java:19]: Please register adapters for `AudioContent` and `VideoContent` in `GsonChatMessageAdapter`:

```java
    private static final Gson GSON = new GsonBuilder()
            .registerTypeAdapter(Content.class, new GsonContentAdapter())
            .registerTypeAdapter(TextContent.class, new GsonContentAdapter())
            .registerTypeAdapter(ImageContent.class, new GsonContentAdapter())
            .create();
```

Would be also great to update `ChatMessageSerializerTest.should_serialize_and_deserialize_chat_message` with `AudioContent` and `VideoContent`.

File: langchain4j-core/src/main/java/dev/langchain4j/data/message/AudioContent.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+import dev.langchain4j.data.audio.Audio;
+
+import java.net.URI;
+import java.util.Objects;
+
+import static dev.langchain4j.data.message.ContentType.AUDIO;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotBlank;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
+
+public class AudioContent implements Content {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/message/AudioContent.java:14]: Please update these comments in `UserMessage`:
`Content can be either TextContent or ImageContent.`

**Comment by langchain4j** [comment]
@glaforge do you plan to update documentation in this PR? I am also wondering if we should rename the "Image Inputs" column [here](https://docs.langchain4j.dev/integrations/language-models/) into "Supported Modalities"

**Comment by glaforge** [comment]
> @glaforge do you plan to update documentation in this PR? I am also wondering if we should rename the "Image Inputs" column [here](https://docs.langchain4j.dev/integrations/language-models/) into "Supported Modalities"

I've updated the table, and explicitly mentioned image, audio, video, pdf support.

File: langchain4j-core/src/main/java/dev/langchain4j/data/pdf/PdfFile.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+import java.net.URI;
+import java.util.Objects;
+
+import static dev.langchain4j.internal.Utils.quoted;
+
+@Experimental
+public class PdfFile {
+    private final URI url;
+    private final String base64Data;
+    private final String mimeType;
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/pdf/PdfFile.java]:
```suggestion
```
No need in mimeType as it is always `application/pdf`, right?

Please remove it from all builders/ctors here and in `PdfFileContent`

File: langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java [hunk 10]
```diff
@@ -132,15 +174,34 @@ private static Part map(TextContent content) {
...
+        return getPart(audio.url(), audio.mimeType(), audio.base64Data());
+    }
+
+    static Part map(VideoContent content) {
+        Video video = content.video();
+        return getPart(video.url(), video.mimeType(), video.base64Data());
+    }
+
+    static Part map(PdfFileContent content) {
+        PdfFile PdfFile = content.pdfFile();
```
Comment (langchain4j) [inline langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/PartsMapper.java]: nit
```suggestion
        PdfFile pdfFile = content.pdfFile();
```

File: docs/docs/integrations/language-models/index.md [hunk 11]
```diff
@@ -4,21 +4,21 @@ hide_title: false
...
+| Provider                                                               | [Streaming](/tutorials/response-streaming) | [Tools](/tutorials/tools) | Supported Modalities     | Local                                                   | Native |
+|------------------------------------------------------------------------|--------------------------------------------|---------------------------|--------------------------|---------------------------------------------------------|--------|
+| [Amazon Bedrock](/integrations/language-models/amazon-bedrock)         |                                            |                           |                          |                                                         |        |
+| [Anthropic](/integrations/language-models/anthropic)                   | ✅                                          | ✅                         | image                    |                                                         | ✅      |
```
Comment (langchain4j) [inline docs/docs/integrations/language-models/index.md]:
```suggestion
| [Anthropic](/integrations/language-models/anthropic)                   | ✅                                          | ✅                         | text, image                    |                                                         | ✅      |
```
Please add `text` to all columns

**Review by ghost — COMMENTED** [review]
@glaforge thank you, good job, almost ready to merge!

**Review by ghost — APPROVED** [review]
@glaforge great job, thank you!

**Pull request merged**